src/NuGet.Clients/NuGet.VisualStudio.Common/NuGetFeedbackDiagnosticFileProvider.cs

```diff
@@ -1,24 +1,110 @@
+
         public IReadOnlyCollection<string> GetFiles()
         {
-            // todo: https://github.com/NuGet/Home/issues/8605
+            string filePath = GetFilePath();
+
+            // See comments on IFeedbackDiagnosticFileProvider.GetFiles
+            ThreadHelper.JoinableTaskFactory.RunAsync(() => WriteToZip(filePath))
+                .PostOnFailure(nameof(NuGetFeedbackDiagnosticFileProvider), nameof(WriteToZip));
 
             return new List<string>()
             {
-                DependencyGraphRestoreUtility.GetDefaultDGSpecFileName()
+                filePath
             };
         }
+
+        public async Task WriteToZip(string filePath)
+        {
+            var telemetry = new TelemetryEvent("feedback");
+            telemetry["IsDebuggerAttached"] = Debugger.IsAttached;
+            bool successful = false;
+            Stopwatch sw = Stopwatch.StartNew();
+            try
+            {
+
+                using (var fileStream = File.Open(filePath, FileMode.Create, FileAccess.Write, FileShare.None))
+                using (var zip = new ZipArchive(fileStream, ZipArchiveMode.Create))
+                {
+                    await AddDgSpecAsync(zip);
+                }
+
+                successful = true;
+            }
+            finally
+            {
+                sw.Stop();
+                telemetry["successful"] = successful;
+                telemetry["duration_ms"] = sw.Elapsed.TotalMilliseconds;
+                TelemetryActivity.EmitTelemetryEvent(telemetry);
+            }
+        }
+
+        private string GetFilePath()
+        {
+            string fileNamePrefix = "nuget." + DateTime.UtcNow.ToString("yyyy-MM-dd.HH-mm-ss");
+            int attempts = 0;
+            for (; ; )
+            {
+                attempts++;
+                var fileName = attempts == 0 ? fileNamePrefix + ".zip" : fileNamePrefix + "." + attempts + ".zip";
+                var fullPath = Path.Combine(Path.GetTempPath(), fileName);
+                if (!File.Exists(fullPath))
+                {
+                    return fullPath;
+                }
+            }
+        }
+
+        private async Task AddDgSpecAsync(ZipArchive zip)
+        {
+            string? tempFile = null;
+            try
+            {
+                if (SolutionManager == null) { return; }
+
+                var dgspec = await DependencyGraphRestoreUtility.GetSolutionRestoreSpec(SolutionManager, context: null);
+                tempFile = Path.GetTempFileName();
+                dgspec.Save(tempFile);
+
+                _ = zip.CreateEntryFromFile(tempFile, "dgspec.json");
+            }
+            catch (Exception exception)
+            {
+                await TelemetryUtility.PostFaultAsync(exception, nameof(NuGetFeedbackDiagnosticFileProvider));
+            }
+            finally
+            {
+                if (tempFile != null)
+                {
+                    File.Delete(tempFile);
```

Comment:
If it's on a temp folder, why NuGet have to delete the file? Would it be better to just keep the file and eventually, let the Operating System delete the file?

Reply:
I have a low opinion of software that doesn't immediately clean up after itself. To me it's a signal about how much the developers care about details and customer experience. I don't care if the operating system will eventually clean up. The oldest file in my own temp directory is from December 2021, making it 7-8 months old (although the next oldest is from May, so "only" 2-3 months old)

Reply:
But deleting the file would affect the time limit for files to being collected?

Reply:
Technically yes, but using the following BenchmarkDotNet test

<details>
<summary>benchmark code</summary>
<p>

```cs
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

BenchmarkRunner.Run<B>();

public class B
{
    [Benchmark]
    public void Delete()
    {
        Test(true);
    }

    [Benchmark]
    public void NoDelete()
    {
        Test(false);
    }

    private void Test(bool delete)
    {
        var tempPath = Path.GetTempPath();
        var tempFilename = "zivkan-" + Path.GetRandomFileName();
        var fullPath = Path.Combine(tempPath, tempFilename);
        using (File.Open(fullPath, FileMode.CreateNew, FileAccess.Write))
        {
        }

        if (delete)
        {
            File.Delete(fullPath);
        }
    }
}
```

</p>
</details>

I get the following results on my local machine with physical SSD:

|   Method |     Mean |    Error |   StdDev |
|--------- |---------:|---------:|---------:|
|   Delete | 254.8 us |  4.10 us |  3.84 us |
| NoDelete | 235.4 us | 12.97 us | 38.25 us |

On an Azure VM with Premium SSD

|   Method |     Mean |    Error |    StdDev |   Median |
|--------- |---------:|---------:|----------:|---------:|
|   Delete | 328.6 us |  6.51 us |  14.82 us | 322.6 us |
| NoDelete | 337.5 us | 54.59 us | 152.18 us | 255.8 us |

On an Azure VM with SSD

|   Method |     Mean |     Error |    StdDev |   Median |
|--------- |---------:|----------:|----------:|---------:|
|   Delete | 322.0 us |   5.04 us |   4.21 us | 321.1 us |
| NoDelete | 704.4 us | 290.70 us | 819.92 us | 259.6 us |

Given that even on cloud machines with virtual disks, delete takes 0.1ms, and the timeout for VS Feedback is 5000ms, I think it's acceptable.